pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,152 @@
+package pxc
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"time"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/users"
+	"github.com/pkg/errors"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileUsers(cr *api.PerconaXtraDBCluster) error {
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	if len(cr.Spec.Users.Secrets) == 0 {
+		return nil
+	}
+	secretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
+		},
+		&secretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get cluster secret '%s'", cr.Spec.SecretsName)
+	}
+	operator := corev1.Pod{}
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      os.Getenv("HOSTNAME"),
+		},
+		&operator,
+	)
+	if err != nil {
+		return errors.Wrap(err, "get operator deployment")
+	}
+
+	for _, secretName := range cr.Spec.Users.Secrets {
+		err = r.handleUsersSecret(secretName, operator, secretObj, cr)
+		if err != nil {
+			log.Error(err, "handle users secret "+secretName)
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) handleUsersSecret(secretName string, operatorPod corev1.Pod, secretObj corev1.Secret, cr *api.PerconaXtraDBCluster) error {
+	containerImage := operatorPod.Spec.Containers[0].Image
+	imagePullSecrets := operatorPod.Spec.ImagePullSecrets
+	usersSecretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      secretName,
+		},
+		&usersSecretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get users secret '%s'", "secret-for-users")
+	}
+
+	newHash := ""
+	if secretData, ok := usersSecretObj.Data["secret.yaml"]; ok {
+		newHash = sha256Hash(secretData)
+	}
+	lastAppliedHash := ""
+	if hash, ok := usersSecretObj.Annotations["last-applied"]; ok {
+		lastAppliedHash = hash
+		if lastAppliedHash == newHash {
+			if usersSecretObj.Annotations["status"] == "succeded" || usersSecretObj.Annotations["status"] == "failed" {
+				return nil
+			}
+		}
+	}
+
+	if len(usersSecretObj.Annotations) == 0 {
+		usersSecretObj.Annotations = make(map[string]string)
+	}
+	usersSecretObj.Annotations["last-applied"] = newHash
```

Comment:
avoid a situation when status was `succeded` for old hash
```suggestion
	usersSecretObj.Annotations["status"] = "applying"
	usersSecretObj.Annotations["last-applied"] = newHash
```
lines 113-119 can be removed

Reply:
done https://github.com/percona/percona-xtradb-cluster-operator/pull/342/commits/3417c02c2c6246ae92274a0c62f72501f6937a3b